pennylane/math/quantum.py

```diff
@@ -181,51 +182,64 @@ def _density_matrix_from_matrix(density_matrix, indices, check_state=False):
             integer number of wires``N``.
         indices (list(int)): List of indices in the considered subsystem.
         check_state (bool): If True, the function will check the state validity (shape and norm).
+        c_dtype (str): Complex floating point precision type.
 
     Returns:
         tensor_like: Density matrix of size ``(2**len(wires), 2**len(wires))``
 
     **Example**
 
     >>> x = np.array([[1, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]])
-    >>> _density_matrix_from_matrix(x, indices=[0])
+    >>> reduce_dm(x, indices=[0])
     [[1.+0.j 0.+0.j]
      [0.+0.j 0.+0.j]]
 
     >>> y = [[0.5, 0, 0.5, 0], [0, 0, 0, 0], [0.5, 0, 0.5, 0], [0, 0, 0, 0]]
-    >>> _density_matrix_from_matrix(y, indices=[0])
+    >>> reduce_dm(y, indices=[0])
     [[0.5+0.j 0.5+0.j]
      [0.5+0.j 0.5+0.j]]
 
-    >>> _density_matrix_from_matrix(y, indices=[1])
+    >>> reduce_dm(y, indices=[1])
     [[1.+0.j 0.+0.j]
      [0.+0.j 0.+0.j]]
 
     >>> z = tf.Variable([[1, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]], dtype=tf.complex128)
-    >>> _density_matrix_from_matrix(x, indices=[1])
+    >>> reduce_dm(x, indices=[1])
     tf.Tensor(
     [[1.+0.j 0.+0.j]
      [0.+0.j 0.+0.j]], shape=(2, 2), dtype=complex128)
 
 
     """
-    dim = density_matrix.shape[0]
-    num_indices = int(np.log2(dim))
+    density_matrix = cast(density_matrix, dtype=c_dtype)
 
     if check_state:
         _check_density_matrix(density_matrix)
 
+    if len(np.shape(density_matrix)) == 2:
+        batch_dim, dim = None, density_matrix.shape[0]
+    else:
+        batch_dim, dim = density_matrix.shape[:2]
+
+    num_indices = int(np.log2(dim))
     consecutive_indices = list(range(num_indices))
 
     # Return the full density matrix if all the wires are given, potentially permuted
     if len(indices) == num_indices:
-        return _permute_dense_matrix(density_matrix, consecutive_indices, indices, None)
+        return _permute_dense_matrix(density_matrix, consecutive_indices, indices, batch_dim)
+
+    if batch_dim is None:
+        density_matrix = qml.math.stack([density_matrix])
 
     # Compute the partial trace
     traced_wires = [x for x in consecutive_indices if x not in indices]
     density_matrix = _partial_trace(density_matrix, traced_wires)
+
+    if batch_dim is None:
+        density_matrix = density_matrix[0]
+
     # Permute the remaining indices of the density matrix
-    return _permute_dense_matrix(density_matrix, sorted(indices), indices, None)
+    return _permute_dense_matrix(density_matrix, sorted(indices), indices, batch_dim)
 
 
 def _partial_trace(density_matrix, indices):

```

Comment:
this function now assumes that it's being passed a batched dm, even if that just means the caller passes a non-batched one wrapped in a list (eg. `dm = [dm]`) right? if so - can the docstring and maybe even the name be updated accordingly? same for similar functions below

Reply:
Yep, that's right. Sure I can update the docstring.